clang/lib/Sema/SemaDeclCXX.cpp

```diff
@@ -18356,23 +18359,32 @@ void Sema::ActOnCXXExitDeclInitializer(Scope *S, Decl *D) {
   if (S && D->isOutOfLine())
     ExitDeclaratorContext(S);
 
-  if (getLangOpts().CPlusPlus23) {
-    // An expression or conversion is 'manifestly constant-evaluated' if it is:
-    // [...]
-    // - the initializer of a variable that is usable in constant expressions or
-    //   has constant initialization.
-    if (auto *VD = dyn_cast<VarDecl>(D);
-        VD && (VD->isUsableInConstantExpressions(Context) ||
-               VD->hasConstantInitialization())) {
-      // An expression or conversion is in an 'immediate function context' if it
-      // is potentially evaluated and either:
-      // [...]
-      // - it is a subexpression of a manifestly constant-evaluated expression
-      //   or conversion.
-      ExprEvalContexts.back().InImmediateFunctionContext = true;
+  if (auto *VD = dyn_cast<VarDecl>(D)) {
+    if (VD->isUsableInConstantExpressions(Context) ||
+        VD->hasConstantInitialization()) {
+      if (getLangOpts().CPlusPlus23) {
+        // An expression or conversion is 'manifestly constant-evaluated' if it
+        // is:
+        // [...]
+        // - the initializer of a variable that is usable in constant
+        // expressions or
+        //   has constant initialization.
+        // An expression or conversion is in an 'immediate function context' if
+        // it is potentially evaluated and either:
+        // [...]
+        // - it is a subexpression of a manifestly constant-evaluated expression
+        //   or conversion.
+        ExprEvalContexts.back().InImmediateFunctionContext = true;
+      }
+    } else {
+      for (auto &DDEntry : MaybeDeviceDeferredDiags)
+        for (auto &DD : DDEntry.second)
+          DeviceDeferredDiags[DDEntry.first].push_back(DD);
```

Comment:
Am I understanding correctly that at this point we know that the variable does not have constexpr initializer, so you're adding all diagnostics for it to DeferredDiags for emission?

Reply:
Yep, that is right. This is the point when we fully parsed and analyzed initializer of a variable so now we know whether it is constexpr or not.